test/e2e/opm_test.go

```diff
@@ -295,43 +307,82 @@ var _ = Describe("opm", func() {
 				td, err := ioutil.TempDir(".", "opm-")
 				Expect(err).NotTo(HaveOccurred())
 				defer os.RemoveAll(td)
 
-				err = bundle.BuildFunc(bundlePaths[i], td, bundleImage+":"+bundleTags[i], containerTool, "", "", "", true)
+				err = bundle.BuildFunc(b.path, td, b.image, containerTool, "", "", "", true)
 				Expect(err).NotTo(HaveOccurred())
 			}
 
 			By("pushing bundles")
-			for _, tag := range bundleTags {
-				err := pushWith(containerTool, bundleImage+":"+tag)
-				Expect(err).NotTo(HaveOccurred())
+			for _, b := range bundles {
+				Expect(pushWith(containerTool, b.image)).NotTo(HaveOccurred())
 			}
 
 			By("building an index")
-			err := buildIndexWith(containerTool, indexImage, bundleImage, bundleTags)
+			indexImage := "quay.io/olmtest/e2e-index:" + rand.String(6)
+			err := buildIndexWith(containerTool, "", indexImage, bundles.images(), registry.ReplacesMode, false)
 			Expect(err).NotTo(HaveOccurred())
 
 			workingDir, err := os.Getwd()
 			Expect(err).NotTo(HaveOccurred())
 			err = os.Remove(workingDir + "/" + bundle.DockerFile)
 			Expect(err).NotTo(HaveOccurred())
 		})
+
+		It("can overwrite existing bundles in an index", func() {
+
+			bundles := bundleLocations{
+				{bundleImage + ":" + rand.String(6), "./testdata/aqua/0.0.1"},
+				{bundleImage + ":" + rand.String(6), "./testdata/aqua/0.0.2"},
+				{bundleImage + ":" + rand.String(6), "./testdata/aqua/1.0.0"},
+				{bundleImage + ":" + rand.String(6), "./testdata/aqua/1.0.1"},
+				{bundleImage + ":" + rand.String(6), "./testdata/aqua/1.0.1-overwrite"},
+			}
+
+			for _, b := range bundles {
+				td, err := ioutil.TempDir(".", "opm-")
+				Expect(err).NotTo(HaveOccurred())
+				defer os.RemoveAll(td)
+
+				err = bundle.BuildFunc(b.path, td, b.image, containerTool, "", "", "", true)
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			By("pushing bundles")
+			for _, b := range bundles {
+				Expect(pushWith(containerTool, b.image)).NotTo(HaveOccurred())
+			}
+
+			indexImage := "quay.io/olmtest/e2e-index:" + rand.String(6)
+			By("adding net-new bundles to an index")
+			err := buildIndexWith(containerTool, "", indexImage, bundles[:4].images(), registry.ReplacesMode, true) // 0.0.1, 0.0.2, 1.0.0, 1.0.1
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pushWith(containerTool, indexImage)).NotTo(HaveOccurred())
+
+			By("failing to overwrite a non-latest bundle")
+			nextIndex := indexImage + "-next"
+			err = buildIndexWith(containerTool, indexImage, nextIndex, bundles[1:2].images(), registry.ReplacesMode, true) // 0.0.2
+			Expect(err).To(HaveOccurred())
+
+			By("failing to overwrite in a non-replace mode")
+			err = buildIndexWith(containerTool, indexImage, nextIndex, bundles[4:].images(), registry.SemVerMode, true) // 1.0.1-overwrite
+			Expect(err).To(HaveOccurred())
+
+			By("overwriting the latest bundle in an index")
+			err = buildIndexWith(containerTool, indexImage, nextIndex, bundles[4:].images(), registry.ReplacesMode, true) // 1.0.1-overwrite
```

Comment:
so this returns a "missing replaces" error for v1.0.1 -> v1.0.0, which I think indicates that there's an issue reconstituting the package during the overwrite procedure.